README.md

```diff
@@ -0,0 +1,116 @@
+Konig Kontext - A globally shared context for Kotlin gRPC microservice architectures
+====================================================================================
+
+Do you have a gRPC microservices architecture? Have you ever needed a value deep in your call stack but the value
+is only known several RPC's ago in the request lifetime? Hate having to update multiple RPC requests/responses in order to pass this value down
+the stack, leading to new code branches and more unit tests? Konig Kontext was developed for this exact frustration. 
+
+See https://github.com/konigsoftware/konig-kontext-example for an example project that implements this library.
+
+Installation
+------------
+
+### Gradle:
+Add the following to your build.gradle.kts:
+```gradle
+implementation("org.konigsoftware:konig-kontext:1.0.0")
+```
+or for Groovy add the following to your build.gradle:
+```gradle
+implementation 'org.konigsoftware:konig-kontext:1.0.0'
+```
+
+Setup
+-----
+
+This library allows you to define and pass around an arbitrary context, in the format of a protobuf message, between all 
+your gRPC microservices. First you need to define the protobuf message format that will be used for the shared context:
+```protobuf
+syntax = "proto3";
+
+package example.shared;
+
+message MyContext {
+  // Add any fields you like here   
+}
+```
+Next there are a few boilerplate items that you need to take care of on both the client and server side.
+
+### Client side setup:
+
+The client side setup just has one step. Simply add the `KonigKontextClientInterceptor` to all of your gRPC clients:
+
+```kotlin
+val myServiceClient = MyServiceCoroutineStub(myManagedChannel).withInterceptors(
+    KonigKontextClientInterceptor()
+)
+
+// Or for a slightly more succinct way you can optionally use the idiomatic helper function:
+val myServiceClient = MyServiceCoroutineStub(myManagedChannel).withKonigKontextInterceptor()
+```
+
+### Server side setup:
+
+On the server side, there are two steps required. First extend the `KonigKontextServer` interface in all of your gRPC service
+implementations and pass in the context protobuf message you created above as the type parameter:
+```kotlin
+class MyService : MyServiceCoroutineImplBase(), KonigKontextServer<MyContext> {
+    // Your service implementation here
+}
+```
+Then when building and running your services, add the `KonigKontextServerInterceptor` with the `addKonigKontextServer` builder function:
+```kotlin
+val myServiceServer = ServerBuilder
+    .forPort(<your_port_here>)
+    .addKonigKontextServer(MyService(), MyContext::class)
```

Comment:
Is there any way you could do is so that you either:

- extend `KonigKontextServer` and then add here normally
..or
- don't extend `KonigKontextServer` but add it here with the context wrapper

?

Reply:
As discussed IRL, I don't think there's a way to do this without extending the `KonigKontextServer`. And this could be done like: 
```kotlin
val myServiceServer = ServerBuilder
    .forPort(<your_port_here>)
    .intercept(KonigKontextCoroutineServerInterceptor(MyContext::class))
    .intercept(KonigKontextGrpcContextServerInterceptor(MyContext::class))
    .addService(MyService(), MyContext::class)
```
But thats a fair bit of boilerplate, and it could be overlooked that you need to add 2 interceptors. So it seems slightly more ergonomic to just have the `addKonigKontextServer` helper function to abstract this boilerplate away?

Reply:
yes agree, and good point about the function enforcing that you have also extended `KonigKontextServer`